CHI_TECH/ChiMath/matrix_operations.cc

```diff
@@ -0,0 +1,512 @@
+    M[1][1] = A[0][0]*A[2][2]*A[3][3] + A[0][2]*A[2][3]*A[3][0] + A[0][3]*A[2][0]*A[3][2] - A[0][0]*A[2][3]*A[3][2] - A[0][2]*A[2][0]*A[3][3] - A[0][3]*A[2][2]*A[3][0];
+    M[1][2] = A[0][0]*A[1][3]*A[3][2] + A[0][2]*A[1][0]*A[3][3] + A[0][3]*A[1][2]*A[3][0] - A[0][0]*A[1][2]*A[3][3] - A[0][2]*A[1][3]*A[3][0] - A[0][3]*A[1][0]*A[3][2];
+    M[1][3] = A[0][0]*A[1][2]*A[2][3] + A[0][2]*A[1][3]*A[2][0] + A[0][3]*A[1][0]*A[2][2] - A[0][0]*A[1][3]*A[2][2] - A[0][2]*A[1][0]*A[2][3] - A[0][3]*A[1][2]*A[2][0];
+
+    M[2][0] = A[1][0]*A[2][1]*A[3][3] + A[1][1]*A[2][3]*A[3][0] + A[1][3]*A[2][0]*A[3][1] - A[1][0]*A[2][3]*A[3][1] - A[1][1]*A[2][0]*A[3][3] - A[1][3]*A[2][1]*A[3][0];
+    M[2][1] = A[0][0]*A[2][3]*A[3][1] + A[0][1]*A[2][0]*A[3][3] + A[0][3]*A[2][1]*A[3][0] - A[0][0]*A[2][1]*A[3][3] - A[0][1]*A[2][3]*A[3][0] - A[0][3]*A[2][0]*A[3][1];
+    M[2][2] = A[0][0]*A[1][1]*A[3][3] + A[0][1]*A[1][3]*A[3][0] + A[0][3]*A[1][0]*A[3][1] - A[0][0]*A[1][3]*A[3][1] - A[0][1]*A[1][0]*A[3][3] - A[0][3]*A[1][1]*A[3][0];
+    M[2][3] = A[0][0]*A[1][3]*A[2][1] + A[0][1]*A[1][0]*A[2][3] + A[0][3]*A[1][1]*A[2][0] - A[0][0]*A[1][1]*A[2][3] - A[0][1]*A[1][3]*A[2][0] - A[0][3]*A[1][0]*A[2][1];
+
+    M[3][0] = A[1][0]*A[2][2]*A[3][1] + A[1][1]*A[2][0]*A[3][2] + A[1][2]*A[2][1]*A[3][0] - A[1][0]*A[2][1]*A[3][2] - A[1][1]*A[2][2]*A[3][0] - A[1][2]*A[2][0]*A[3][1];
+    M[3][1] = A[0][0]*A[2][1]*A[3][2] + A[0][1]*A[2][2]*A[3][0] + A[0][2]*A[2][0]*A[3][1] - A[0][0]*A[2][2]*A[3][1] - A[0][1]*A[2][0]*A[3][2] - A[0][2]*A[2][1]*A[3][0];
+    M[3][2] = A[0][0]*A[1][2]*A[3][1] + A[0][1]*A[1][0]*A[3][2] + A[0][2]*A[1][1]*A[3][0] - A[0][0]*A[1][1]*A[3][2] - A[0][1]*A[1][2]*A[3][0] - A[0][2]*A[1][0]*A[3][1];
+    M[3][3] = A[0][0]*A[1][1]*A[2][2] + A[0][1]*A[1][2]*A[2][0] + A[0][2]*A[1][0]*A[2][1] - A[0][0]*A[1][2]*A[2][1] - A[0][1]*A[1][0]*A[2][2] - A[0][2]*A[1][1]*A[2][0];
+    Scale(M,f);
+  }
+  else
+    M = InverseGEPivoting(A);
+
+  return M;
+}
+
+
+//######################################################### Power iteration
+/** Performs power iteration to obtain the fundamental eigen mode. The
+ * eigen-value of the fundamental mode is return whilst the eigen-vector
+ * is return via reference.*/
+double ChiMath::Fundamental_EigValVec_PowerIteration(
+  const MatDbl &A, VecDbl &e_vec, int max_it, double tol)
+{
+  // Local Variables
+  unsigned int n = A.size();
+  int it_counter = 0;
+  VecDbl y(n,1.);
+  double lambda0 = 0.;
+
+  // Perform initial iteration outside of loop
+  VecDbl Ay = MatMul(A, y);
+  double lambda = Dot( y, Ay );
+  y = VecMul( Ay, 1./Vec2Norm(Ay) );
+  if (lambda < 0.)
+    Scale(y, -1.0);
+
+  // Perform convergence loop
+  bool converged = false;
+  while(!converged && it_counter < max_it)
+  {
+    // Update old eigenvalue
+    lambda0 = lambda;
+    // Calculate new eigenvalue/eigenvector
+    Ay = MatMul(A, y);
+    lambda = Dot( y, Ay );
+    y = VecMul( Ay, 1./Vec2Norm(Ay) );
+    if (lambda < 0.)
```

Comment:
The sign flipping could be done after the convergence has been reached, just before returning -- although this is more a simplification of the algorithm rather than a significant performance boost.

Reply:
Done, although now I needed another fabs.